### ISSUE TITLE
Doesn't work with Google Play Services 8.3 and 8.4

### ISSUE THREAD

**Issue by antygravity** [robolectric/robolectric#2215]
With version 8.3(
    compile 'com.google.android.gms:play-services-auth:8.3.0'
    compile 'com.google.android.gms:play-services-maps:8.3.0'
    compile 'com.google.android.gms:play-services-location:8.3.0'
),  
compileSdkVersion 23
    buildToolsVersion '23.0.2'

 got error:
java.lang.NullPointerException: Expecting a valid IBinder
	at com.google.android.gms.common.internal.zzx.zzb(Unknown Source)
	at com.google.android.gms.common.internal.zzx.zzb(Unknown Source)
	at com.google.android.gms.common.internal.zzj$zze.onServiceConnected(Unknown Source)
	at com.google.android.gms.common.internal.zzj$zze.onServiceConnected(Unknown Source)
	at com.google.android.gms.common.internal.zzm$zzb$zza.onServiceConnected(Unknown Source)
	at com.google.android.gms.common.internal.zzm$zzb$zza.onServiceConnected(Unknown Source)
	at org.robolectric.shadows.ShadowApplication$2.run(ShadowApplication.java:310)
	at org.robolectric.util.Scheduler$ScheduledRunnable.run(Scheduler.java:269)
	at org.robolectric.util.Scheduler.runOneTask(Scheduler.java:179)
	at org.robolectric.util.Scheduler.advanceTo(Scheduler.java:160)
	at org.robolectric.util.Scheduler.advanceBy(Scheduler.java:143)
	at org.robolectric.util.Scheduler.unPause(Scheduler.java:54)
	at org.robolectric.shadows.ShadowLooper.unPause(ShadowLooper.java:266)
	at org.robolectric.shadows.ShadowLooper.runPaused(ShadowLooper.java:306)
	at org.robolectric.shadows.CoreShadowsAdapter$2.runPaused(CoreShadowsAdapter.java:45)
	at org.robolectric.util.ComponentController.invokeWhilePaused(ComponentController.java:69)
	at org.robolectric.util.ActivityController.start(ActivityController.java:143)
	at org.robolectric.util.ActivityController.setup(ActivityController.java:210)
	at org.robolectric.Robolectric.setupActivity(Robolectric.java:46)

with version: 8.4
java.lang.NullPointerException
	at com.google.android.gms.common.zze.zzi(Unknown Source)
	at com.google.android.gms.common.zze.zzi(Unknown Source)
	at com.google.android.gms.common.zze.zzd(Unknown Source)
	at com.google.android.gms.common.zze.zzd(Unknown Source)
	at com.google.android.gms.common.zzc.isGooglePlayServicesAvailable(Unknown Source)
	at com.google.android.gms.common.zzc.isGooglePlayServicesAvailable(Unknown Source)
	at com.google.android.gms.common.api.internal.zzh$zzb.zzpt(Unknown Source)
	at com.google.android.gms.common.api.internal.zzh$zzb.zzpt(Unknown Source)
	at com.google.android.gms.common.api.internal.zzh$zzf.run(Unknown Source)
	at com.google.android.gms.common.api.internal.zzh$zzf.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at com.google.android.gms.internal.zznl.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)

My test:

@RunWith(RobolectricGradleTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 21)
public class MainActivityTest {

@Before
    public void setUp() {
       ShadowGooglePlayServicesUtil.setIsGooglePlayServicesAvailable(ConnectionResult.SUCCESS);
    }

    @Test
    public void shouldShowLoginButtons() throws Exception {

        Activity activity = Robolectric.setupActivity(MainActivity.class);
        TextView facebookLoginButton = (TextView) activity.findViewById(R.id.login_button_facebook);
        assertNotNull(facebookLoginButton);
}
}
Thanks,
Marcin

**Comment by KioKrofovitch** [comment]
I am also seeing this exact error, though I am only pulling in location services:
`compile 'com.google.android.gms:play-services-location:8.3.0'`

**Comment by guitcastro** [comment]
I am also seeing this exact error =/

**Comment by guitcastro** [comment]
It's seems that [GooglePlayServicesUtil.isGooglePlayServicesAvailable](https://developers.google.com/android/reference/com/google/android/gms/common/GooglePlayServicesUtil.html#isGooglePlayServicesAvailable(android.content.Context))  war replace by [GoogleApiAvailability.isGooglePlayServicesAvailable](https://developers.google.com/android/reference/com/google/android/gms/common/GoogleApiAvailability).

I tried to create a custom shadow, but did not worked.

```
/**
 * Shadow for {@link com.google.android.gms.common.GoogleApiAvailability}.
 */
@Implements(GoogleApiAvailability.class)
public class ShadowGoogleApiAvailability {
    private static int availabilityCode = ConnectionResult.SERVICE_MISSING;


    @Implementation
    public int isGooglePlayServicesAvailable(Context context) {
        return availabilityCode;
    }

    public static void setIsGooglePlayServicesAvailable(int availabilityCode) {
        ShadowGoogleApiAvailability.availabilityCode = availabilityCode;
    }
}
```
Calling  `GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(RuntimeEnvironment.application);` does call my Shadow implementation, although when I run my test it does no call the Shadow. No ideia why =/

It fail with the same exception: 
```
java.lang.NullPointerException
	at com.google.android.gms.common.zze.zzi(Unknown Source)
	at com.google.android.gms.common.zze.zzi(Unknown Source)
	at com.google.android.gms.common.zze.zzd(Unknown Source)
	at com.google.android.gms.common.zze.zzd(Unknown Source)
	at com.google.android.gms.common.zzc.isGooglePlayServicesAvailable(Unknown Source)
	at com.google.android.gms.common.zzc.isGooglePlayServicesAvailable(Unknown Source)
	at com.google.android.gms.common.GoogleApiAvailability.isGooglePlayServicesAvailable(Unknown Source)
	at com.google.android.gms.common.GoogleApiAvailability.isGooglePlayServicesAvailable(Unknown Source)
 ```

**Comment by guitcastro** [comment]
Ok, further more investigations:

The `zzc` class is an private class from Google Play Services that actual implement `isGooglePlayServicesAvailable ` and the GoogleApiAvailability extends from that class. 

This class is used internally, so if I implement the shadow for the 'zzc' class it "worked", but I get another error. At this point I give up creating shadows for internals Google Play Services Apis. 

= /

**Comment by diegotori** [comment]
I remember a couple of months ago, I was able to create a ShadowGoogleApiClientBuilder (https://github.com/diegotori/robolectric/blob/shadows-play-services-improvements/robolectric-shadows/shadows-play-services/src/main/java/org/robolectric/shadows/gms/ShadowGoogleApiClientBuilder.java) which returned a fake GoogleApiClient implementation, back when it was an interface. Now that the library is poised to support the latest Google Play Services version, it should be a lot easier to create a ShadowGoogleApiClient + Builder that returns the shadow since its now a base abstract class. 

Hopefully later this week, I might take another crack at fixing that up and PR'ing it soon.

**Comment by balajir12** [comment]
Thanks @diegotori for looking into this. It is greatly inconvenient to ignore tests due to the issue.

**Comment by forrestbice** [comment]
@diegotori any chance there's a PR out to address this / what's the status?

**Comment by diegotori** [comment]
Currently I'm trying to close out the related changes made in my topic branch: https://github.com/diegotori/robolectric/tree/shadows_play_services_marshmallow_support

I was trying to model the ShadowGoogleApiClient.Builder based on how it was done on for ShadowNotification: https://github.com/robolectric/robolectric/blob/master/robolectric-shadows/shadows-core/src/main/resources/org/robolectric/shadows/ShadowNotification.java.vm

However, last time I ran the ShadowGoogleApiClient.Builder tests that I created, some of them failed due to the fact that I was using the reflection utility methods to set them directly on the real object, which caused NPEs when calling build() due to the fact that it didn't instantiate the real object properly. I could use a bit of help getting through that issue, because that way, the real object would not NPE when calling Shadow.directlyOn to inject the Builder's params directly.

On the bright side, GoogleApiAvailability has been minimally shadowed in that branch (ShadowGoogleApiAvailability).

Bottom line, I'll attempt tonight to at least PR ShadowGoogleApiAvailability, since that one's passing all of its tests. However, I'm going to need help on trying to instantiate GoogleApiClient.Builder's RealObject with the Context passed into it, so that reflection actually ends up working.

**Comment by forrestbice** [comment]
Awesome, thanks for the the update @diegotori! I'll keep an eye out for the `ShadowGoogleApiAvailability` PR

**Comment by diegotori** [comment]
Didn't have time last night to submit the PR, but I will be able to tonight and/or all-day tomorrow.

**Comment by diegotori** [comment]
Submitted Pull Request for ShadowGoogleApiAvailability. Hopefully it gets approved before they mint 3.1, which I'm hearing could happen any day now.

**Comment by forrestbice** [comment]
Great news; kudos and thank you @diegotori!

**Comment by asuraphel** [comment]
@diegotori when will your PR be merged? What is preventing it?

**Comment by diegotori** [comment]
@asuraphel I have to re-sync it with master tonight since I last did so this past Sunday. Once that's done and it passes unit testing, the rest is up to the Robolectric team to approve and merge into master.

**Comment by asuraphel** [comment]
@diegotori Thanks!

**Comment by diegotori** [comment]
Pull request #2319 was merged a few days ago, so that tackles part of the issue. GoogleApiClient + Builder is still a WIP on a separate branch in my fork. Just gotta re-sync it and finish it.

**Comment by AndreSand** [comment]
Same issue Robolectric 3.0 not working with Google Play Services 8.4. Solved with work around add code below; the work around will skip Google Play Services code.

   public static boolean isRoboUnitTest() {
        return "robolectric".equals(Build.FINGERPRINT);
    }

if (!isRoboUnitTest()) {
// Google Play Services code....
//...... ...
}

My unit tests pass now 👍

**Comment by forrestbice** [comment]
What's the best way to track releases, etc.? I can't seem to find information on when 3.1 might be released. Apologies in advance if this is located somewhere obvious and I'm just missing it

**Comment by shenoybrs** [comment]
@diegotori  i took the latest robolectric drop and your shadow fix for the playservice is working fine with 8.4 playservice ..
